wpe/src/main/java/com/wpe/wpeview/WPECookieManager.java

```diff
@@ -0,0 +1,110 @@
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+package com.wpe.wpeview;
+
+import android.annotation.SuppressLint;
+import android.os.Handler;
+
+import androidx.annotation.Keep;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+public final class WPECookieManager {
+
+    private enum NativeCookieAcceptPolicy {
+        AcceptAlways(0),
+        AcceptNever(1),
+        AcceptNoThirdParty(2);
+
+        private final int value;
+
+        NativeCookieAcceptPolicy(int value) { this.value = value; }
+
+        public int getValue() { return value; }
+    }
+
+    private static final WPECookieManager instance = new WPECookieManager();
+
+    // Keep these local because fetching them is async in wpe
+    private boolean acceptCookie = true;
+    private boolean acceptThirdpartyCookie = false;
+
+    private WPECookieManager() {}
+
+    public static @NonNull WPECookieManager getInstance() { return instance; }
+
+    public boolean getAcceptCookie() { return acceptCookie; }
+
+    public void setAcceptCookie(boolean accept) {
+        if (acceptCookie == accept)
+            return;
+
+        acceptCookie = accept;
+
+        NativeCookieAcceptPolicy policy = NativeCookieAcceptPolicy.AcceptNever;
+        if (accept) {
+            if (!acceptThirdpartyCookie) {
+                policy = NativeCookieAcceptPolicy.AcceptNoThirdParty;
+            } else {
+                policy = NativeCookieAcceptPolicy.AcceptAlways;
+            }
+        }
+
+        nativeSetCookieAcceptPolicy(policy.getValue());
+    }
+
+    public boolean getAcceptThirdPartyCookies() { return acceptThirdpartyCookie; }
+
+    public void setAcceptThirdPartyCookies(boolean accept) {
+        if (acceptThirdpartyCookie == accept)
+            return;
+
+        acceptThirdpartyCookie = accept;
+
+        // Update wpe only we accept cookies in general
+        if (acceptCookie) {
+            NativeCookieAcceptPolicy policy = acceptThirdpartyCookie ? NativeCookieAcceptPolicy.AcceptAlways
+                                                                     : NativeCookieAcceptPolicy.AcceptNoThirdParty;
+            nativeSetCookieAcceptPolicy(policy.getValue());
+        }
+    }
+
+    public void removeAllCookies(@Nullable Callback callback) {
+        if (callback != null) {
+            nativeRemoveAllCookies(callback);
+        }
+    }
+
+    // CookieManagerCallback that calls a Callback on its original thread
+    @FunctionalInterface
+    public interface Callback {
+        Handler handler = new Handler();
```

Comment:
When a variable is immutable and can be accessed from different threads, it should be `final`.
Normally the static analyzer should issue a warning but I suppose that here it has not detected that `handler` member may be accessed from different threads because there is no concrete implementation of the interface.

Any concrete implementation of this interface through lambda or class, will take the final modifier into account.

As a general matter it's a good practice the set `final` all variables that you know will be immutable, it brings some interesting properties for multithreading safety:

https://www.javamex.com/tutorials/synchronization_final.shtml

Reply:
My understanding is that all fields in interfaces are basically by default already "public static final".
Also Android Studio tells me that
"
Modifier 'final' is redundant for interface fields
"

Reply:
Effectively, you're right. I forgot about this trap with member variables in interfaces. It's even considered as an anti-pattern because it causes a lot of confusion (like the one I just did ;) ).

With `handler` being `static` it means that all implementations of `Callback` will use the same `handler` that will be initiliaze when the class loader loads the class in memory. So all callbacks will be called in the same thread, which will be the first thread that has called this.

I understand that what you want to obtain is that the callback is called in the same thread as the one which called `removeAllCookies()` with passing this callback ?

In this case, we may separate the callback behavior in two like, for example:
```
@FunctionalInterface
public interface Callback {
    void onResult(boolean result);
}

private static final class CallbackHolder {
    private final Handler handler = new Handler();
    private final Callback callback;

    public CallbackHolder(Callback cb) {
        callback = cb;
    }

    @Keep
    public void commitResult(boolean result) {
        handler.post(() -> callback.onResult(result));
        }
}
```

This way in `removeAllCookies()` instead of calling `nativeRemoveAllCookies(callback);` you can call: `nativeRemoveAllCookies(new CallbackHolder(callback));` and the native code would be exactly the same (except for  the class name that would be `CallbackHolder`).

Reply:
So in the end we are almost back to the very original interface that I had in the beginning I think